expat/tests/xmltest.sh

```diff
@@ -1,147 +1,279 @@
 #! /usr/bin/env bash
+#
+# EXPAT TEST SCRIPT FOR W3C XML TEST SUITE
+#
+# This script can be used to exercise Expat against the w3c.org xml test
+# suite, available from:
+#
+#   <http://www.w3.org/XML/Test/xmlts20020606.zip>
+#
+# The script lists  all test cases where Expat shows  a discrepancy from
+# the  expected result.   Test  cases where  only  the canonical  output
+# differs  are prefixed  with  "Output  differs:", and  a  diff file  is
+# generated in the appropriate subdirectory under $OUTDIR.
+#
+# If there  are output files provided,  the script will use  output from
+# xmlwf and compare the desired output  against it.  However, one has to
+# take into account that the canonical output produced by xmlwf conforms
+# to an older definition of canonical XML and does not generate notation
+# declarations.
 
-#   EXPAT TEST SCRIPT FOR W3C XML TEST SUITE
+#page
+#### global settings
 
-# This script can be used to exercise Expat against the
-# w3c.org xml test suite, available from
-# http://www.w3.org/XML/Test/xmlts20020606.zip.
+shopt -s nullglob
 
-# To run this script, first set XMLWF below so that xmlwf can be
-# found, then set the output directory with OUTPUT.
+#page
+#### global variables
 
-# The script lists all test cases where Expat shows a discrepancy
-# from the expected result. Test cases where only the canonical
-# output differs are prefixed with "Output differs:", and a diff file
-# is generated in the appropriate subdirectory under $OUTPUT.
+# The environment  variable EXPAT_ABS_SRCDIR is  set in the  Makefile to
+# the source directory.  The  environment variable EXPAT_ABS_BUILDDIR is
+# set in the Makefile to the build directory.
+#
+declare -r EXPAT_ABS_SRCDIR="${EXPAT_ABS_SRCDIR:?'environment variable EXPAT_ABS_SRCDIR is not set'}"
+declare -r EXPAT_ABS_BUILDDIR="${EXPAT_ABS_BUILDDIR:?'environment variable EXPAT_ABS_BUILDDIR is not set'}"
 
-# If there are output files provided, the script will use
-# output from xmlwf and compare the desired output against it.
-# However, one has to take into account that the canonical output
-# produced by xmlwf conforms to an older definition of canonical XML
-# and does not generate notation declarations.
+declare -r XMLWF="${1:-${EXPAT_ABS_BUILDDIR}/xmlwf/xmlwf}"
+# declare -r XMLWF=/usr/local/bin/xmlwf
 
-shopt -s nullglob
+# Unicode-aware diff utility.  Requires Python version 3.
+#
+declare -r DIFF="${EXPAT_ABS_SRCDIR}/tests/udiffer.py"
+
+declare -r XML_BASE_SRCDIR="$EXPAT_ABS_BUILDDIR"/tests/xmlconf
+declare -r XML_BASE_DSTDIR="$EXPAT_ABS_BUILDDIR"/tests/xmltest-output
```

Comment:
Style note: My vote for `${foo}` over `$foo` everywhere. Here it would be consistent with `declare -r XMLWF="${1:-${EXPAT_ABS_BUILDDIR}/xmlwf/xmlwf}"` above.